src/ray/core_worker/transport/direct_task_transport.cc

```diff
@@ -76,11 +76,39 @@ void CoreWorkerDirectTaskSubmitter::OnWorkerIdle(
     // because this is the only place tasks are removed from it.
     if (queue_entry->second.empty()) {
       task_queues_.erase(queue_entry);
+      RAY_LOG(DEBUG) << "Task queue empty, canceling lease request";
+      CancelWorkerLeaseIfNeeded(scheduling_key);
     }
   }
   RequestNewWorkerIfNeeded(scheduling_key);
 }
 
+void CoreWorkerDirectTaskSubmitter::CancelWorkerLeaseIfNeeded(
+    const SchedulingKey &scheduling_key) {
+  auto queue_entry = task_queues_.find(scheduling_key);
+  if (queue_entry != task_queues_.end()) {
+    // There are still pending tasks, so let the worker lease request succeed.
+    return;
+  }
+
+  auto it = pending_lease_requests_.find(scheduling_key);
+  if (it != pending_lease_requests_.end()) {
+    auto &lease_client = it->second.first;
+    auto &lease_id = it->second.second;
+    RAY_LOG(DEBUG) << "Canceling lease request " << lease_id;
+    RAY_UNUSED(lease_client->CancelWorkerLease(
+        lease_id, [this, lease_id, scheduling_key](
+                      const Status &status, const rpc::CancelWorkerLeaseReply &reply) {
+          absl::MutexLock lock(&mu_);
+          if (!status.ok() || reply.success()) {
+            RequestNewWorkerIfNeeded(scheduling_key);
+          } else {
+            CancelWorkerLeaseIfNeeded(scheduling_key);
```

Comment:
Is it possible cancellation failed, but we don't need to retry? For example, if the lease was granted before we could cancel it. Or should that be considered a successful cancellation?

Reply:
Oh oops, I did mean to count a granted lease as a successful cancellation in the raylet. I'll update it. Unfortunately, I can't think of a good way to test that.